File: 1417. Reformat The String/1417.go
package leetcode

func reformat(s string) string {
	letter := make([]string, 0, len(s)/2)
	dight := make([]string, 0, len(s)/2)

	for _, v := range s{
		if v>='0' && v<='9'{
			dight = append(dight, string(v))
		}else{
			letter = append(letter, string(v))
		}
	}

	tmp := len(letter)-len(dight)
	if tmp > 1 || tmp < -1{
		return ""
	}

	res := ""
	if len(letter)>len(dight){
		for i,v := range dight{
			res += letter[i]
			res += v
		}
		res += letter[len(letter)-1]
	}else if len(letter)<len(dight){
		for i,v := range letter{
			res += dight[i]
			res += v
		}
		res += dight[len(dight)-1]
	}else{
		for i,v := range letter{
			res += dight[i]
			res += v
		}
	}

	return res
}

File: 1417. Reformat The String/1417.c


char * reformat(char * s){
    int slen = strlen(s);
    char* digit = (char*)malloc(slen);
    memset(digit, 0, slen);
    int digitlen = 0;
    char* letter = (char*)malloc(slen);
    int letterlen = 0;
    memset(letter, 0, slen);

    char* res = (char*)malloc(slen+1);
    memset(res, 0, slen+1);

    while (*s != 0){
        if (*s>='0' && *s<='9'){
            *(digit+digitlen) = *s;
            digitlen++;
        } else{
            *(letter+letterlen) = *s;
            letterlen++;
        }
        s ++;
    }

    int tmp = digitlen - letterlen;
    if (tmp>1 || tmp<-1){
        return "";
    }

    if(digitlen>letterlen){
        for(int i=0; i<letterlen; i++){
            *(res+i*2) = *(digit+i);
            *(res+i*2+1)=*(letter+i);
        }
        *(res+slen-1) = *(digit+digitlen-1);
    }else if(digitlen<letterlen){
        for(int i=0; i<digitlen; i++){
            *(res+i*2) = *(letter+i);
            *(res+i*2+1)=*(digit+i);
        }
        *(res+slen-1) = *(letter+letterlen-1);
    }else{
        for(int i=0; i<digitlen; i++){
            *(res+i*2) = *(letter+i);
            *(res+i*2+1)=*(digit+i);
        }
    }
    return res;
}